Return a SwiftArrayEmptyBufferHandler if loc points to an EmptyArray Symbol.

rastogishubham opened the pull request:
If we have a ValueObjectSP who's storage_location points to a `S_swiftEmptyArrayStorage` symbol, return a `SwiftArrayEmptyBufferHandler`

kastiglione commented on lldb/source/Plugins/Language/Swift/SwiftArray.cpp:
```diff
@@ -460,6 +461,21 @@ SwiftArrayBufferHandler::CreateBufferHandler(ValueObject &static_valobj) {
+    // a SwiftArrayEmptyBufferHandler.
+    if (auto *symbol = addr.CalculateSymbolContextSymbol()) {
+      auto mangledName = symbol->GetMangled().GetMangledName().GetStringRef();
+      if (mangledName == "$ss19__EmptyArrayStorageCN") {
```
should this be `_swiftEmptyArrayStorage`?

rastogishubham replied:
I am checking the mangled name:

```
> xcrun swift-demangle ss19__EmptyArrayStorageCN 
$ss19__EmptyArrayStorageCN ---> type metadata for Swift.__EmptyArrayStorage
```

kastiglione replied:
When I inspect an empty array, I see:

```
(lldb) v -R -T -d no-dynamic-values
(Swift.Array<main.P>) x = {
  (Swift._ArrayBuffer<main.P>) _buffer = {
    (Swift._BridgeStorage<Swift.__ContiguousArrayStorageBase>) _storage = {
      (Builtin.BridgeObject) rawValue = 0x00000001f52cd120 _swiftEmptyArrayStorage
    }
  }
}
```

I don't see where `$ss19__EmptyArrayStorageCN` comes into play. Also the PR description mentions `_swiftEmptyArrayStorage` but the changed code does not.

adrian-prantl replied:
This mangled name is for the *type metadata* for `Swift.__EmptyArrayStorage`.
1. You probably want to match against `Swift.__EmptyArrayStorage` itself.
2. Instead of depending on the mangling it would be better to use the helpers in SwiftDemangling.h to make sure you have the expected demangle tree for a class with the right name in STDLIB_MODULENAME.

rastogishubham committed:
Return a SwiftArrayEmptyBufferHandler if loc points to an EmptyArray Symbol.

If we have a ValueObjectSP who's storage_location points to a
S_swiftEmptyArrayStorage symbol, return a SwiftArrayEmptyBufferHandler

rdar://141180827

kastiglione commented on lldb/test/API/lang/swift/swift_empty_arr/TestSwiftGlobalEmptyArray.py:
```diff
@@ -0,0 +1,17 @@
+        target, process, thread, breakpoint1 = lldbutil.run_to_source_breakpoint(
+            self, "break here", filespec
+        )
+        self.expect("p x", substrs=["([a.P]) 0 values {}"])
```
for tests we use `frame var` or `expr` to be explicit whether it's use expression evaluation or not.

adrian-prantl replied:
Even better would be
```
x = self.frame().FindVariable("x")
lldbutil.check_variable(self, x, summary="0 values")
```

adrian-prantl replied:
or `self.assertEqual(x.GetSummary(), "0 values")`

adrian-prantl commented on lldb/test/API/lang/swift/swift_empty_arr/TestSwiftGlobalEmptyArray.py:
```diff
@@ -0,0 +1,17 @@
+
+        self.build()
+        filespec = lldb.SBFileSpec("main.swift")
+        target, process, thread, breakpoint1 = lldbutil.run_to_source_breakpoint(
```
```suggestion
        lldbutil.run_to_source_breakpoint(
```

adrian-prantl commented on lldb/test/API/lang/swift/swift_empty_arr/TestSwiftGlobalEmptyArray.py:
```diff
@@ -0,0 +1,17 @@
+        self.build()
+        filespec = lldb.SBFileSpec("main.swift")
+        target, process, thread, breakpoint1 = lldbutil.run_to_source_breakpoint(
+            self, "break here", filespec
```
```suggestion
            self, "break here", lldb.SBFileSpec("main.swift")
```

rastogishubham commented:
As discussed with @adrian-prantl yesterday, we tried really hard to get a reliable testcase for this issue but were not able to create one
